src/transport/SessionMessageDelegate.h

```diff
@@ -0,0 +1,59 @@
+/*
+ *    Copyright (c) 2021 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <transport/SessionHandle.h>
+
+namespace chip {
+
+/**
+ * @brief
+ *   This class provides a skeleton for the callback functions. The functions will be
+ *   called by SecureSssionMgrBase object on specific events. If the user of SessionManager
+ *   is interested in receiving these callbacks, they can specialize this class and handle
+ *   each trigger in their implementation of this class.
+ */
+class DLL_EXPORT SessionMessageDelegate
+{
+public:
+    virtual ~SessionMessageDelegate() {}
+
+    enum class DuplicateMessage : uint8_t
+    {
+        Yes,
+        No,
+    };
+
+    /**
+     * @brief
+     *   Called when a new message is received. The function must internally release the
+     *   msgBuf after processing it.
+     *
+     * @param packetHeader  The message header
+     * @param payloadHeader The payload header
+     * @param session       The handle to the secure session
+     * @param source        The sender's address
+     * @param isDuplicate   The message is a duplicate of previously received message
+     * @param msgBuf        The received message
+     */
+    virtual void OnMessageReceived(const PacketHeader & packetHeader, const PayloadHeader & payloadHeader, SessionHandle session,
+                                   const Transport::PeerAddress & source, DuplicateMessage isDuplicate,
+                                   System::PacketBufferHandle && msgBuf)
+    {}
```

Comment:
would it make sense for this to be pure virtual? Otherwise we allow inheiting SessionMessageDelegate without implementing the only virtual function in it, which does not seem as useful.